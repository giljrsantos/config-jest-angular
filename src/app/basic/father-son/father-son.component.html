<h5>Filho</h5>
<code>{{ client | json }}</code>

<br /><br />

<button
  *ngIf="client"
  class="btn btn-outline-primary"
  (click)="onChange(5)"
  data-test="btn-id"
>
  Mudar ID
</button>
<button
  *ngIf="client"
  class="btn btn-outline-danger"
  (click)="onDelete()"
  data-test="btn-delete"
>
  Apagar Cliente
</button>
