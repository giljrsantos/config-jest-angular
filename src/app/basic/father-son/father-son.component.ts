import {
  Component,
  EventEmitter,
  Input,
  Output,
} from '@angular/core';

import { Client } from '../interface';

@Component({
  selector: 'app-father-son',
  templateUrl: './father-son.component.html',
  styleUrls: ['./father-son.component.scss'],
})
export class FatherSonComponent {
  @Input() client?: Client;
  @Output() onDeleteClient = new EventEmitter();
  @Output() onClientUpdated = new EventEmitter<Client>();

  onDelete() {
    this.client = undefined;
    this.onDeleteClient.emit();
  }

  onChange(id: number) {
    if (!this.client) return;
    //this.client.id = newId; n√£o usar
    this.client = {
      ...this.client,
      id,
    };
    this.onClientUpdated.emit({ ...this.client });
  }
}
