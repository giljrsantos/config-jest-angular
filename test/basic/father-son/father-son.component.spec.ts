import { ComponentFixture, TestBed } from '@angular/core/testing';

import { FatherSonComponent } from '../../../src/app/basic/father-son/father-son.component';

describe('FatherSonComponent', () => {
  let component: FatherSonComponent;
  let fixture: ComponentFixture<FatherSonComponent>;
  let compiled: HTMLElement;

  beforeEach(() => {
    TestBed.configureTestingModule({
      declarations: [FatherSonComponent]
    }).compileComponents()
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(FatherSonComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
    compiled = fixture.nativeElement;
    jest.clearAllMocks();
  });


  it('deveria criar', () => {
    expect(component).toBeTruthy();
  });

  it('deve corresponder ao snapshot', () => {
    expect(compiled).toMatchSnapshot();
  });

  it('não deve aparecer botões se não tiver cliente', () => {
    const buttons = compiled.querySelectorAll('button');
    expect(buttons.length).toBe(0);
  });
  it('deve aparecer 2 botões se tiver cliente', () => {
    component.client = { id: 1, name: 'Juan' };
    fixture.detectChanges();

    const buttons = compiled.querySelectorAll('button');
    expect(buttons.length).toBe(2);

  });
  it('deve aparecer 2 botões se tiver cliente no snapshot', () => {
    component.client = { id: 1, name: 'Juan' };
    fixture.detectChanges();

    expect(compiled).toMatchSnapshot();

  });

  it('deve emitir onDeleteClient com o botão delete', () => {
    component.client = { id: 1, name: 'Juan' };
    fixture.detectChanges();

    jest.spyOn(component.onDeleteClient, 'emit');

    const btnDelete = compiled.querySelector('[data-test=btn-delete]')
    btnDelete?.dispatchEvent(new Event('click'));
    expect(component.onDeleteClient.emit).toHaveBeenCalled();

  });
  it('deve emitir onClientUpdated com o botão "Mudar ID"', () => {
    component.client = { id: 1, name: 'Juan' };
    fixture.detectChanges();

    jest.spyOn(component.onClientUpdated, 'emit');

    const btnChangeId = compiled.querySelector('[data-test=btn-id]')
    btnChangeId?.dispatchEvent(new Event('click'));
    expect(component.onClientUpdated.emit).toHaveBeenCalledWith({
      id: 5,
      name: 'Juan'
    });

  });

  it('deve lançar onChangeClient com o ID especificado SE houver um cliente', () => {

    jest.spyOn(component.onClientUpdated, 'emit');
    component.onChange(10);
    expect(component.onClientUpdated.emit).not.toHaveBeenCalled();

    component.client = { id: 1, name: 'Juan' };
    fixture.detectChanges();
    component.onChange(10);

    expect(component.onClientUpdated.emit).toHaveBeenCalledWith({
      id: 10,
      name: 'Juan'
    })


  });

});
