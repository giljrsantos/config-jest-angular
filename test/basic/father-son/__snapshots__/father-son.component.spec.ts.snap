// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FatherSonComponent deve aparecer 2 bot√µes se tiver cliente no snapshot 1`] = `
<div
  id="root4"
>
  <h5>
    Filho
  </h5>
  <code>
    {
  "id": 1,
  "name": "Juan"
}
  </code>
  <br />
  <br />
  <button
    class="btn btn-outline-primary"
    data-test="btn-id"
  >
     Mudar ID

  </button>
  
  <button
    class="btn btn-outline-danger"
    data-test="btn-delete"
  >
     Apagar Cliente

  </button>
  
</div>
`;

exports[`FatherSonComponent deve corresponder ao snapshot 1`] = `
<div
  id="root1"
>
  <h5>
    Filho
  </h5>
  <code>
    
  </code>
  <br />
  <br />
  
  
</div>
`;
